modules/openid_connect/app/services/openid_connect/user_tokens/exchange_service.rb

```diff
@@ -65,6 +67,11 @@ def call(audience)
         # A second reason is that at least Keycloak (an IDP we implement against), offers broken
         # refresh tokens after token exchange (see https://github.com/keycloak/keycloak/issues/37016)
         token = store_exchanged_token(audience:, access_token:, refresh_token: nil, expires_in:)
+
+        storage = Storages::Storage.all.find { |s| s.audience == audience }
```

Comment:
:red_circle: I think this code should be in the `FetchService`, otherwise it would only be called on token exchange and many times we do not perform a token exchange.

:yellow_circle: I think we should not hardwire this generic piece of authentication code to _storages_ specifically. We now have a direct dependency from OIDC to storages.

Reply:
Done. Moved to `FetchService` and decoupled using event pubsub.